contrib/babelfishpg_tsql/sql/sys_views.sql

```diff
@@ -49,11 +64,12 @@ select
   , CAST(null as integer) as history_table_id
   , CAST(0 as sys.bit) as is_remote_data_archive_enabled
   , CAST(0 as sys.bit) as is_external
-from pg_class t inner join sys.schemas sch on t.relnamespace = sch.schema_id
-where t.relpersistence in ('p', 'u', 't')
+from pg_class t 
+where t.relnamespace in (select schema_id from sys.schemas)
```

Comment:
It is possible that optimizer will simply treat this in as semi-join, unless we have only a few schemas. 
Since this is an performance optimization, I would like to know more about the effect of this change.

Reply:
Experimented with the queries described in the JIRA [BABEL-3423](https://jira.rds.a2z.com/browse/BABEL-3423) using data attached in the JIRA.

This change had no (noticeable) effect on the SSMS generated queries.

However, for the first query below execution time dropped from 8 minutes 54 seconds to 5 seconds.

The second query below initially ran for hours and had to kill it. I didn't test it just with this one change, but given that the two share common structure, I think the second query was also dramatically improved. 
With all the changes in this PR the second query runs in 26 seconds (down from over 5 hours).

```sql
select 
  tbl.name AS [Name]
  , clmns.name
from sys.tables AS tbl
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id
```

```sql
select 
    tbl.name AS [Name]
    , clmns.name
    , usrt.name as tname
from sys.tables AS tbl
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id
inner join sys.types as usrt on usrt.user_type_id = clmns.user_type_id;
```

This doc gives more context:
https://quip-amazon.com/kSRrAAk6bxJR#ZeC9AAQzDNP

Reply:
1. Could you provide query plan changes in your test env? It could help us to understand the problem and solution better.
2. Not for this specific PR, but we need a test framework that can detect performance regressions.

Reply:
Yes, will do.

Reply:
Here are the query plans

**Query 1**

```sql
select 
  tbl.name AS [Name]
  , clmns.name
from sys.tables AS tbl
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id
```

**Plan before the changes**

```
Query Text: select  
  tbl.name AS Name 
  , clmns.name 
from sys.tables AS tbl 
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id
Nested Loop  (cost=235.70..1054.71 rows=1 width=64)
  Join Filter: ((t.oid)::integer = (c.oid)::integer)
  ->  Nested Loop  (cost=220.78..439.59 rows=1 width=68)
        ->  Nested Loop  (cost=0.15..13.26 rows=1 width=4)
              ->  Seq Scan on pg_namespace base  (cost=0.00..1.07 rows=1 width=68)
                    Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[])))
              ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext  (cost=0.15..8.17 rows=1 width=64)
                    Index Cond: (nspname = base.nspname)
                    Filter: ((dbid)::oid = '5'::oid)
        ->  Bitmap Heap Scan on pg_class t  (cost=220.63..426.17 rows=17 width=72)
              Recheck Cond: (relnamespace = base.oid)
              Filter: (has_table_privilege(oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER'::text COLLATE "default") AND (relkind = 'r'::"char") AND (relpersistence = ANY ('{p,u,t}'::"char"[])) AND (NOT is_table_type(oid)))
              ->  Bitmap Index Scan on pg_class_relname_nsp_index  (cost=0.00..220.63 rows=378 width=0)
                    Index Cond: (relnamespace = base.oid)
  ->  Nested Loop  (cost=14.92..612.59 rows=201 width=68)
        ->  Nested Loop  (cost=14.91..608.57 rows=201 width=72)
              ->  Nested Loop  (cost=14.91..604.55 rows=201 width=76)
                    ->  Nested Loop  (cost=14.62..577.39 rows=201 width=72)
                          Join Filter: has_column_privilege(((quote_ident((s.nspname)::text) || '.'::text COLLATE "default") || quote_ident((c.relname)::text)), (a.attname)::text, 'SELECT,INSERT,UPDATE,REFERENCES'::text COLLATE "default")
                          ->  Hash Join  (cost=14.33..182.58 rows=102 width=132)
                                Hash Cond: (c.relnamespace = s.oid)
                                ->  Seq Scan on pg_class c  (cost=0.00..164.43 rows=1023 width=72)
                                      Filter: (relkind = ANY ('{r,v,m,f,p}'::"char"[]))
                                ->  Hash  (cost=14.32..14.32 rows=1 width=68)
                                      ->  Seq Scan on pg_namespace s  (cost=13.25..14.32 rows=1 width=68)
                                            Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND ((hashed SubPlan 1) OR (nspname = 'sys'::pg_catalog.name)))
                                            SubPlan 1
                                              ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                    ->  Seq Scan on pg_namespace base_1  (cost=0.00..1.06 rows=1 width=68)
                                                          Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                    ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_1  (cost=0.15..8.17 rows=1 width=64)
                                                          Index Cond: (nspname = base_1.nspname)
                                                          Filter: ((dbid)::oid = '5'::oid)
                          ->  Index Scan using pg_attribute_relid_attnum_index on pg_attribute a  (cost=0.29..3.69 rows=6 width=78)
                                Index Cond: ((attrelid = c.oid) AND (attnum > 0))
                                Filter: (NOT attisdropped)
                    ->  Memoize  (cost=0.29..0.32 rows=1 width=8)
                          Cache Key: a.atttypid
                          Cache Mode: logical
                          ->  Index Scan using pg_type_oid_index on pg_type t_1  (cost=0.28..0.31 rows=1 width=8)
                                Index Cond: (oid = a.atttypid)
              ->  Function Scan on translate_pg_type_to_tsql tsql_type_name  (cost=0.00..0.01 rows=1 width=0)
        ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name  (cost=0.00..0.01 rows=1 width=0)
```

**Plan with this PR changes**

```
Query Text: select  
  tbl.name AS Name 
  , clmns.name 
from sys.tables AS tbl 
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id
Nested Loop  (cost=1898.80..2696.55 rows=1 width=64)
  Join Filter: ((t.oid)::integer = (c.oid)::integer)
  ->  Nested Loop  (cost=14.92..607.65 rows=203 width=68)
        ->  Nested Loop  (cost=14.91..603.59 rows=203 width=72)
              ->  Nested Loop  (cost=14.91..599.52 rows=203 width=76)
                    ->  Nested Loop  (cost=14.62..572.93 rows=203 width=72)
                          Join Filter: has_column_privilege(((quote_ident((s.nspname)::text) || '.'::text COLLATE "default") || quote_ident((c.relname)::text)), (a.attname)::text, 'SELECT,INSERT,UPDATE,REFERENCES'::text COLLATE "default")
                          ->  Hash Join  (cost=14.33..182.49 rows=102 width=132)
                                Hash Cond: (c.relnamespace = s.oid)
                                ->  Seq Scan on pg_class c  (cost=0.00..164.34 rows=1022 width=72)
                                      Filter: (relkind = ANY ('{r,v,m,f,p}'::"char"[]))
                                ->  Hash  (cost=14.32..14.32 rows=1 width=68)
                                      ->  Seq Scan on pg_namespace s  (cost=13.25..14.32 rows=1 width=68)
                                            Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND ((hashed SubPlan 2) OR (nspname = 'sys'::pg_catalog.name)))
                                            SubPlan 2
                                              ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                    ->  Seq Scan on pg_namespace base_2  (cost=0.00..1.06 rows=1 width=68)
                                                          Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                    ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_2  (cost=0.15..8.17 rows=1 width=64)
                                                          Index Cond: (nspname = base_2.nspname)
                                                          Filter: ((dbid)::oid = '5'::oid)
                          ->  Index Scan using pg_attribute_relid_attnum_index on pg_attribute a  (cost=0.29..3.65 rows=6 width=78)
                                Index Cond: ((attrelid = c.oid) AND (attnum > 0))
                                Filter: (NOT attisdropped)
                    ->  Memoize  (cost=0.29..0.32 rows=1 width=8)
                          Cache Key: a.atttypid
                          Cache Mode: logical
                          ->  Index Scan using pg_type_oid_index on pg_type t_1  (cost=0.28..0.31 rows=1 width=8)
                                Index Cond: (oid = a.atttypid)
              ->  Function Scan on translate_pg_type_to_tsql tsql_type_name  (cost=0.00..0.01 rows=1 width=0)
        ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name  (cost=0.00..0.01 rows=1 width=0)
  ->  Materialize  (cost=1883.89..2076.72 rows=4 width=68)
        ->  Hash Semi Join  (cost=1883.89..2076.70 rows=4 width=68)
              Hash Cond: (t.relnamespace = base.oid)
              ->  Seq Scan on pg_class t  (cost=1870.63..2063.28 rows=43 width=72)
                    Filter: ((NOT (hashed SubPlan 1)) AND has_schema_privilege(relnamespace, 'USAGE'::text COLLATE "default") AND has_table_privilege(oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER'::text COLLATE "default") AND (relkind = 'r'::"char") AND (relpersistence = ANY ('{p,u,t}'::"char"[])))
                    SubPlan 1
                      ->  Hash Join  (cost=187.68..1858.18 rows=4978 width=4)
                            Hash Cond: (dep.objid = pt.typrelid)
                            ->  Seq Scan on pg_depend dep  (cost=0.00..1488.71 rows=4244 width=4)
                                  Filter: (deptype = 'i'::"char")
                            ->  Hash  (cost=187.39..187.39 rows=23 width=8)
                                  ->  Nested Loop  (cost=13.54..187.39 rows=23 width=8)
                                        ->  Hash Semi Join  (cost=13.26..101.29 rows=114 width=4)
                                              Hash Cond: (pt.typnamespace = base_1.oid)
                                              ->  Seq Scan on pg_type pt  (cost=0.00..84.06 rows=1028 width=8)
                                                    Filter: (typtype = 'c'::"char")
                                              ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                                                    ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                          ->  Seq Scan on pg_namespace base_1  (cost=0.00..1.06 rows=1 width=68)
                                                                Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                          ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_1  (cost=0.15..8.17 rows=1 width=64)
                                                                Index Cond: (nspname = base_1.nspname)
                                                                Filter: ((dbid)::oid = '5'::oid)
                                        ->  Index Scan using pg_class_oid_index on pg_class pc  (cost=0.28..0.76 rows=1 width=4)
                                              Index Cond: (oid = pt.typrelid)
                                              Filter: (relkind = 'r'::"char")
              ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                    ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                          ->  Seq Scan on pg_namespace base  (cost=0.00..1.06 rows=1 width=68)
                                Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                          ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext  (cost=0.15..8.17 rows=1 width=64)
                                Index Cond: (nspname = base.nspname)
                                Filter: ((dbid)::oid = '5'::oid)
```

**Query 2**

```sql
select 
    tbl.name AS [Name]
    , clmns.name
    , usrt.name as tname
from sys.tables AS tbl
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id
inner join sys.types as usrt on usrt.user_type_id = clmns.user_type_id;
```

**Plan before the changes**

```
Query Text: select  
    tbl.name AS [Name] 
    , clmns.name 
    , usrt.name as tname 
from sys.tables AS tbl 
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id 
inner join sys.types as usrt on usrt.user_type_id = clmns.user_type_id
Nested Loop  (cost=236.79..1363.48 rows=1 width=96)
  Join Filter: (t_1.oid = "*SELECT* 1".user_type_id)
  ->  Nested Loop  (cost=235.70..1054.70 rows=1 width=136)
        Join Filter: ((t.oid)::integer = (c.oid)::integer)
        ->  Nested Loop  (cost=220.78..439.59 rows=1 width=68)
              ->  Nested Loop  (cost=0.15..13.26 rows=1 width=4)
                    ->  Seq Scan on pg_namespace base  (cost=0.00..1.07 rows=1 width=68)
                          Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[])))
                    ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext  (cost=0.15..8.17 rows=1 width=64)
                          Index Cond: (nspname = base.nspname)
                          Filter: ((dbid)::oid = '5'::oid)
              ->  Bitmap Heap Scan on pg_class t  (cost=220.63..426.17 rows=17 width=72)
                    Recheck Cond: (relnamespace = base.oid)
                    Filter: (has_table_privilege(oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER'::text COLLATE "default") AND (relkind = 'r'::"char") AND (relpersistence = ANY ('{p,u,t}'::"char"[])) AND (NOT is_table_type(oid)))
                    ->  Bitmap Index Scan on pg_class_relname_nsp_index  (cost=0.00..220.63 rows=378 width=0)
                          Index Cond: (relnamespace = base.oid)
        ->  Nested Loop  (cost=14.92..612.59 rows=201 width=76)
              ->  Nested Loop  (cost=14.91..608.57 rows=201 width=80)
                    ->  Nested Loop  (cost=14.91..604.55 rows=201 width=80)
                          ->  Nested Loop  (cost=14.62..577.39 rows=201 width=72)
                                Join Filter: has_column_privilege(((quote_ident((s.nspname)::text) || '.'::text COLLATE "default") || quote_ident((c.relname)::text)), (a.attname)::text, 'SELECT,INSERT,UPDATE,REFERENCES'::text COLLATE "default")
                                ->  Hash Join  (cost=14.33..182.58 rows=102 width=132)
                                      Hash Cond: (c.relnamespace = s.oid)
                                      ->  Seq Scan on pg_class c  (cost=0.00..164.43 rows=1023 width=72)
                                            Filter: (relkind = ANY ('{r,v,m,f,p}'::"char"[]))
                                      ->  Hash  (cost=14.32..14.32 rows=1 width=68)
                                            ->  Seq Scan on pg_namespace s  (cost=13.25..14.32 rows=1 width=68)
                                                  Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND ((hashed SubPlan 1) OR (nspname = 'sys'::pg_catalog.name)))
                                                  SubPlan 1
                                                    ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                          ->  Seq Scan on pg_namespace base_2  (cost=0.00..1.06 rows=1 width=68)
                                                                Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                          ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_2  (cost=0.15..8.17 rows=1 width=64)
                                                                Index Cond: (nspname = base_2.nspname)
                                                                Filter: ((dbid)::oid = '5'::oid)
                                ->  Index Scan using pg_attribute_relid_attnum_index on pg_attribute a  (cost=0.29..3.69 rows=6 width=78)
                                      Index Cond: ((attrelid = c.oid) AND (attnum > 0))
                                      Filter: (NOT attisdropped)
                          ->  Memoize  (cost=0.29..0.32 rows=1 width=8)
                                Cache Key: a.atttypid
                                Cache Mode: logical
                                ->  Index Scan using pg_type_oid_index on pg_type t_1  (cost=0.28..0.31 rows=1 width=8)
                                      Index Cond: (oid = a.atttypid)
                    ->  Function Scan on translate_pg_type_to_tsql tsql_type_name  (cost=0.00..0.01 rows=1 width=0)
              ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name  (cost=0.00..0.01 rows=1 width=0)
  ->  Append  (cost=1.09..306.57 rows=176 width=36)
        ->  Subquery Scan on "*SELECT* 1"  (cost=1.09..146.90 rows=175 width=36)
              ->  Nested Loop  (cost=1.09..145.15 rows=175 width=146)
                    ->  Hash Join  (cost=1.08..141.65 rows=175 width=4)
                          Hash Cond: (t_2.typnamespace = s_1.oid)
                          ->  Seq Scan on pg_type t_2  (cost=0.00..137.53 rows=786 width=12)
                                Filter: pg_type_is_visible(oid)
                          ->  Hash  (cost=1.06..1.06 rows=2 width=4)
                                ->  Seq Scan on pg_namespace s_1  (cost=0.00..1.06 rows=2 width=4)
                                      Filter: ((nspname = 'pg_catalog'::pg_catalog.name) OR (nspname = 'sys'::pg_catalog.name))
                    ->  Function Scan on translate_pg_type_to_tsql tsql_type_name_1  (cost=0.00..0.01 rows=1 width=32)
                          Filter: (tsql_type_name IS NOT NULL)
        ->  Subquery Scan on "*SELECT* 2"  (cost=39.39..158.79 rows=1 width=36)
              ->  Nested Loop  (cost=39.39..158.78 rows=1 width=146)
                    ->  Nested Loop  (cost=39.14..158.51 rows=1 width=72)
                          ->  Nested Loop  (cost=39.14..158.49 rows=1 width=76)
                                ->  Nested Loop  (cost=39.14..158.46 rows=1 width=76)
                                      ->  Nested Loop  (cost=0.15..14.34 rows=1 width=8)
                                            ->  Nested Loop  (cost=0.00..2.15 rows=1 width=72)
                                                  Join Filter: (s_2.oid = base_1.oid)
                                                  ->  Seq Scan on pg_namespace base_1  (cost=0.00..1.06 rows=1 width=68)
                                                        Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                  ->  Seq Scan on pg_namespace s_2  (cost=0.00..1.04 rows=4 width=4)
                                            ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_1  (cost=0.15..8.17 rows=1 width=64)
                                                  Index Cond: (nspname = base_1.nspname)
                                                  Filter: ((dbid)::oid = '5'::oid)
                                      ->  Bitmap Heap Scan on pg_type t_3  (cost=38.99..143.23 rows=90 width=84)
                                            Recheck Cond: (typnamespace = s_2.oid)
                                            Filter: ((typtype = 'd'::"char") OR is_table_type(typrelid))
                                            ->  Bitmap Index Scan on pg_type_typname_nsp_index  (cost=0.00..38.97 rows=262 width=0)
                                                  Index Cond: (typnamespace = s_2.oid)
                                ->  Function Scan on translate_pg_type_to_tsql tsql_type_name_2  (cost=0.00..0.01 rows=1 width=0)
                                      Filter: (tsql_type_name IS NULL)
                          ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name_1  (cost=0.00..0.01 rows=1 width=0)
                    ->  Function Scan on is_table_type is_tbl_type  (cost=0.25..0.26 rows=1 width=0)
```

**Plan with this PR changes**

```
Query Text: select  
    tbl.name AS [Name] 
    , clmns.name 
    , usrt.name as tname 
from sys.tables AS tbl 
inner join sys.all_columns as clmns on clmns.object_id = tbl.object_id 
inner join sys.types as usrt on usrt.user_type_id = clmns.user_type_id
Nested Loop  (cost=2058.70..6085.83 rows=1 width=96)
  Join Filter: (t_1.oid = t_2.oid)
  ->  Nested Loop  (cost=1898.80..2696.54 rows=1 width=136)
        Join Filter: ((t.oid)::integer = (c.oid)::integer)
        ->  Nested Loop  (cost=14.92..607.65 rows=203 width=76)
              ->  Nested Loop  (cost=14.91..603.59 rows=203 width=80)
                    ->  Nested Loop  (cost=14.91..599.52 rows=203 width=80)
                          ->  Nested Loop  (cost=14.62..572.93 rows=203 width=72)
                                Join Filter: has_column_privilege(((quote_ident((s.nspname)::text) || '.'::text COLLATE "default") || quote_ident((c.relname)::text)), (a.attname)::text, 'SELECT,INSERT,UPDATE,REFERENCES'::text COLLATE "default")
                                ->  Hash Join  (cost=14.33..182.49 rows=102 width=132)
                                      Hash Cond: (c.relnamespace = s.oid)
                                      ->  Seq Scan on pg_class c  (cost=0.00..164.34 rows=1022 width=72)
                                            Filter: (relkind = ANY ('{r,v,m,f,p}'::"char"[]))
                                      ->  Hash  (cost=14.32..14.32 rows=1 width=68)
                                            ->  Seq Scan on pg_namespace s  (cost=13.25..14.32 rows=1 width=68)
                                                  Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND ((hashed SubPlan 2) OR (nspname = 'sys'::pg_catalog.name)))
                                                  SubPlan 2
                                                    ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                          ->  Seq Scan on pg_namespace base_4  (cost=0.00..1.06 rows=1 width=68)
                                                                Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                          ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_4  (cost=0.15..8.17 rows=1 width=64)
                                                                Index Cond: (nspname = base_4.nspname)
                                                                Filter: ((dbid)::oid = '5'::oid)
                                ->  Index Scan using pg_attribute_relid_attnum_index on pg_attribute a  (cost=0.29..3.65 rows=6 width=78)
                                      Index Cond: ((attrelid = c.oid) AND (attnum > 0))
                                      Filter: (NOT attisdropped)
                          ->  Memoize  (cost=0.29..0.32 rows=1 width=8)
                                Cache Key: a.atttypid
                                Cache Mode: logical
                                ->  Index Scan using pg_type_oid_index on pg_type t_1  (cost=0.28..0.31 rows=1 width=8)
                                      Index Cond: (oid = a.atttypid)
                    ->  Function Scan on translate_pg_type_to_tsql tsql_type_name  (cost=0.00..0.01 rows=1 width=0)
              ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name  (cost=0.00..0.01 rows=1 width=0)
        ->  Materialize  (cost=1883.89..2076.72 rows=4 width=68)
              ->  Hash Semi Join  (cost=1883.89..2076.70 rows=4 width=68)
                    Hash Cond: (t.relnamespace = base.oid)
                    ->  Seq Scan on pg_class t  (cost=1870.63..2063.28 rows=43 width=72)
                          Filter: ((NOT (hashed SubPlan 1)) AND has_schema_privilege(relnamespace, 'USAGE'::text COLLATE "default") AND has_table_privilege(oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER'::text COLLATE "default") AND (relkind = 'r'::"char") AND (relpersistence = ANY ('{p,u,t}'::"char"[])))
                          SubPlan 1
                            ->  Hash Join  (cost=187.68..1858.18 rows=4978 width=4)
                                  Hash Cond: (dep_1.objid = pt_1.typrelid)
                                  ->  Seq Scan on pg_depend dep_1  (cost=0.00..1488.71 rows=4244 width=4)
                                        Filter: (deptype = 'i'::"char")
                                  ->  Hash  (cost=187.39..187.39 rows=23 width=8)
                                        ->  Nested Loop  (cost=13.54..187.39 rows=23 width=8)
                                              ->  Hash Semi Join  (cost=13.26..101.29 rows=114 width=4)
                                                    Hash Cond: (pt_1.typnamespace = base_3.oid)
                                                    ->  Seq Scan on pg_type pt_1  (cost=0.00..84.06 rows=1028 width=8)
                                                          Filter: (typtype = 'c'::"char")
                                                    ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                                                          ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                                ->  Seq Scan on pg_namespace base_3  (cost=0.00..1.06 rows=1 width=68)
                                                                      Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                                ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_3  (cost=0.15..8.17 rows=1 width=64)
                                                                      Index Cond: (nspname = base_3.nspname)
                                                                      Filter: ((dbid)::oid = '5'::oid)
                                              ->  Index Scan using pg_class_oid_index on pg_class pc_1  (cost=0.28..0.76 rows=1 width=4)
                                                    Index Cond: (oid = pt_1.typrelid)
                                                    Filter: (relkind = 'r'::"char")
                    ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                          ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                ->  Seq Scan on pg_namespace base  (cost=0.00..1.06 rows=1 width=68)
                                      Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext  (cost=0.15..8.17 rows=1 width=64)
                                      Index Cond: (nspname = base.nspname)
                                      Filter: ((dbid)::oid = '5'::oid)
  ->  Append  (cost=159.90..3354.07 rows=1565 width=146)
        CTE type_code_list
          ->  HashAggregate  (cost=15.00..17.00 rows=200 width=64)
                Group Key: babelfish_typecode_list.pg_typname, babelfish_typecode_list.tsql_typname
                ->  Function Scan on babelfish_typecode_list  (cost=0.00..10.00 rows=1000 width=64)
        ->  Nested Loop  (cost=142.89..162.29 rows=15 width=146)
              ->  Function Scan on name default_collation_name  (cost=0.01..0.01 rows=1 width=64)
              ->  Nested Loop Left Join  (cost=142.89..150.73 rows=15 width=111)
                    ->  Hash Join  (cost=142.60..147.50 rows=15 width=51)
                          Hash Cond: ((ti.pg_type_name)::text = t_2.typname)
                          ->  CTE Scan on type_code_list ti  (cost=0.00..4.00 rows=199 width=64)
                                Filter: (tsql_type_name IS NOT NULL)
                          ->  Hash  (cost=140.45..140.45 rows=172 width=83)
                                ->  Hash Join  (cost=1.08..140.45 rows=172 width=83)
                                      Hash Cond: (t_2.typnamespace = s_1.oid)
                                      ->  Seq Scan on pg_type t_2  (cost=0.00..136.38 rows=775 width=83)
                                            Filter: pg_type_is_visible(oid)
                                      ->  Hash  (cost=1.06..1.06 rows=2 width=4)
                                            ->  Seq Scan on pg_namespace s_1  (cost=0.00..1.06 rows=2 width=4)
                                                  Filter: ((nspname = 'pg_catalog'::pg_catalog.name) OR (nspname = 'sys'::pg_catalog.name))
                    ->  Memoize  (cost=0.29..0.61 rows=1 width=68)
                          Cache Key: t_2.typcollation
                          Cache Mode: logical
                          ->  Index Scan using pg_collation_oid_index on pg_collation c_1  (cost=0.28..0.60 rows=1 width=68)
                                Index Cond: (oid = t_2.typcollation)
        ->  Hash Right Join  (cost=1892.86..3151.29 rows=1550 width=146)
              Hash Cond: (pt.typrelid = t_3.typrelid)
              Filter: ((t_3.typtype = 'd'::"char") OR (pt.typrelid IS NOT NULL))
              ->  Merge Join  (cost=1784.81..1823.33 rows=4978 width=4)
                    Merge Cond: (pt.typrelid = dep.objid)
                    ->  Sort  (cost=105.18..105.47 rows=114 width=4)
                          Sort Key: pt.typrelid
                          ->  Hash Semi Join  (cost=13.26..101.29 rows=114 width=4)
                                Hash Cond: (pt.typnamespace = base_2.oid)
                                ->  Seq Scan on pg_type pt  (cost=0.00..84.06 rows=1028 width=8)
                                      Filter: (typtype = 'c'::"char")
                                ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                                      ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                            ->  Seq Scan on pg_namespace base_2  (cost=0.00..1.06 rows=1 width=68)
                                                  Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                            ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_2  (cost=0.15..8.17 rows=1 width=64)
                                                  Index Cond: (nspname = base_2.nspname)
                                                  Filter: ((dbid)::oid = '5'::oid)
                    ->  Sort  (cost=1679.62..1680.75 rows=452 width=8)
                          Sort Key: dep.objid
                          ->  Hash Join  (cost=159.82..1659.69 rows=452 width=8)
                                Hash Cond: (dep.objid = pc.oid)
                                ->  Seq Scan on pg_depend dep  (cost=0.00..1488.71 rows=4244 width=4)
                                      Filter: (deptype = 'i'::"char")
                                ->  Hash  (cost=150.19..150.19 rows=771 width=4)
                                      ->  Seq Scan on pg_class pc  (cost=0.00..150.19 rows=771 width=4)
                                            Filter: (relkind = 'r'::"char")
              ->  Hash  (cost=108.04..108.04 rows=1 width=248)
                    ->  Nested Loop  (cost=103.12..108.04 rows=1 width=248)
                          ->  Nested Loop  (cost=103.12..108.02 rows=1 width=216)
                                ->  Function Scan on name default_collation_name_1  (cost=0.01..0.01 rows=1 width=64)
                                ->  Nested Loop Left Join  (cost=103.11..107.99 rows=1 width=152)
                                      ->  Hash Right Join  (cost=102.83..107.59 rows=1 width=92)
                                            Hash Cond: (ti_1.pg_type_name = t_3.typname)
                                            Filter: (ti_1.tsql_type_name IS NULL)
                                            ->  CTE Scan on type_code_list ti_1  (cost=0.00..4.00 rows=200 width=64)
                                            ->  Hash  (cost=102.81..102.81 rows=2 width=92)
                                                  ->  Hash Join  (cost=13.26..102.81 rows=2 width=92)
                                                        Hash Cond: (t_3.typnamespace = base_1.oid)
                                                        ->  Seq Scan on pg_type t_3  (cost=0.00..78.25 rows=2325 width=92)
                                                        ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                                                              ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                                    ->  Seq Scan on pg_namespace base_1  (cost=0.00..1.06 rows=1 width=68)
                                                                          Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                                    ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_1  (cost=0.15..8.17 rows=1 width=64)
                                                                          Index Cond: (nspname = base_1.nspname)
                                                                          Filter: ((dbid)::oid = '5'::oid)
                                      ->  Index Scan using pg_collation_oid_index on pg_collation c_2  (cost=0.28..0.40 rows=1 width=68)
                                            Index Cond: (oid = t_3.typcollation)
                          ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name_1  (cost=0.00..0.01 rows=1 width=32)
```

**Query 3**

```sql
select 
	tbl.name AS [Name]
	, tbl.schema_id AS [Schema]
	, CAST(ISNULL((select top 1 1 from all_columns as clmns join sys.types as usrt on usrt.user_type_id = clmns.user_type_id where clmns.object_id = tbl.object_id and usrt.name = N'xml'), 0) AS bit) AS [HasXmlData]
from sys.tables AS tbl; 
```

**Plan before the changes**

```
Query Text: select  
 tbl.name AS [Name] 
 , tbl.schema_id AS [Schema] 
 , CAST(ISNULL((select top 1 1 from all_columns as clmns join sys.types as usrt on usrt.user_type_id = clmns.user_type_id where clmns.object_id = tbl.object_id and usrt.name = N'xml'), 0) AS bit) AS [HasXmlData] 
from sys.tables AS tbl
Nested Loop  (cost=220.78..956.87 rows=1 width=37)
  ->  Nested Loop  (cost=0.15..13.26 rows=1 width=4)
        ->  Seq Scan on pg_namespace base  (cost=0.00..1.07 rows=1 width=68)
              Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[])))
        ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext  (cost=0.15..8.17 rows=1 width=64)
              Index Cond: (nspname = base.nspname)
              Filter: ((dbid)::oid = '5'::oid)
  ->  Bitmap Heap Scan on pg_class t  (cost=220.63..426.17 rows=17 width=72)
        Recheck Cond: (relnamespace = base.oid)
        Filter: (has_table_privilege(oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER'::text COLLATE "default") AND (relkind = 'r'::"char") AND (relpersistence = ANY ('{p,u,t}'::"char"[])) AND (NOT is_table_type(oid)))
        ->  Bitmap Index Scan on pg_class_relname_nsp_index  (cost=0.00..220.63 rows=378 width=0)
              Index Cond: (relnamespace = base.oid)
  SubPlan 2
    ->  Limit  (cost=14.91..517.27 rows=1 width=4)
          ->  Nested Loop  (cost=14.91..517.27 rows=1 width=4)
                Join Filter: (t_1.oid = "*SELECT* 1".user_type_id)
                ->  Nested Loop  (cost=13.82..207.66 rows=1 width=8)
                      ->  Nested Loop  (cost=13.82..207.64 rows=1 width=12)
                            ->  Nested Loop  (cost=13.82..207.62 rows=1 width=12)
                                  ->  Nested Loop  (cost=13.54..207.31 rows=1 width=4)
                                        Join Filter: has_column_privilege(((quote_ident((s.nspname)::text) || '.'::text COLLATE "default") || quote_ident((c.relname)::text)), (a.attname)::text, 'SELECT,INSERT,UPDATE,REFERENCES'::text COLLATE "default")
                                        ->  Nested Loop  (cost=13.25..188.21 rows=1 width=132)
                                              Join Filter: (c.relnamespace = s.oid)
                                              ->  Seq Scan on pg_class c  (cost=0.00..173.88 rows=1 width=72)
                                                    Filter: (((oid)::integer = (t.oid)::integer) AND (relkind = ANY ('{r,v,m,f,p}'::"char"[])))
                                              ->  Seq Scan on pg_namespace s  (cost=13.25..14.32 rows=1 width=68)
                                                    Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND ((hashed SubPlan 1) OR (nspname = 'sys'::pg_catalog.name)))
                                                    SubPlan 1
                                                      ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                            ->  Seq Scan on pg_namespace base_1  (cost=0.00..1.06 rows=1 width=68)
                                                                  Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                            ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_1  (cost=0.15..8.17 rows=1 width=64)
                                                                  Index Cond: (nspname = base_1.nspname)
                                                                  Filter: ((dbid)::oid = '5'::oid)
                                        ->  Index Scan using pg_attribute_relid_attnum_index on pg_attribute a  (cost=0.29..18.92 rows=6 width=78)
                                              Index Cond: ((attrelid = c.oid) AND (attnum > 0))
                                              Filter: (NOT attisdropped)
                                  ->  Index Scan using pg_type_oid_index on pg_type t_1  (cost=0.28..0.31 rows=1 width=8)
                                        Index Cond: (oid = a.atttypid)
                            ->  Function Scan on translate_pg_type_to_tsql tsql_type_name  (cost=0.00..0.01 rows=1 width=0)
                      ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name  (cost=0.00..0.01 rows=1 width=0)
                ->  Append  (cost=1.09..307.40 rows=176 width=4)
                      ->  Subquery Scan on "*SELECT* 1"  (cost=1.09..147.34 rows=175 width=4)
                            ->  Nested Loop  (cost=1.09..145.59 rows=175 width=146)
                                  ->  Hash Join  (cost=1.08..141.65 rows=175 width=4)
                                        Hash Cond: (t_2.typnamespace = s_1.oid)
                                        ->  Seq Scan on pg_type t_2  (cost=0.00..137.53 rows=786 width=12)
                                              Filter: pg_type_is_visible(oid)
                                        ->  Hash  (cost=1.06..1.06 rows=2 width=4)
                                              ->  Seq Scan on pg_namespace s_1  (cost=0.00..1.06 rows=2 width=4)
                                                    Filter: ((nspname = 'pg_catalog'::pg_catalog.name) OR (nspname = 'sys'::pg_catalog.name))
                                  ->  Function Scan on translate_pg_type_to_tsql tsql_type_name_1  (cost=0.00..0.02 rows=1 width=0)
                                        Filter: ((tsql_type_name IS NOT NULL) AND (tsql_type_name = 'xml'::text))
                      ->  Subquery Scan on "*SELECT* 2"  (cost=39.37..159.19 rows=1 width=4)
                            ->  Nested Loop  (cost=39.37..159.18 rows=1 width=146)
                                  ->  Nested Loop  (cost=39.12..158.91 rows=1 width=8)
                                        ->  Nested Loop  (cost=39.12..158.88 rows=1 width=12)
                                              ->  Nested Loop  (cost=39.11..158.86 rows=1 width=12)
                                                    ->  Nested Loop  (cost=38.97..146.68 rows=1 width=76)
                                                          ->  Nested Loop  (cost=0.00..2.15 rows=1 width=72)
                                                                Join Filter: (s_2.oid = base_2.oid)
                                                                ->  Seq Scan on pg_namespace base_2  (cost=0.00..1.06 rows=1 width=68)
                                                                      Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                                ->  Seq Scan on pg_namespace s_2  (cost=0.00..1.04 rows=4 width=4)
                                                          ->  Bitmap Heap Scan on pg_type t_3  (cost=38.97..144.52 rows=1 width=20)
                                                                Recheck Cond: (typnamespace = s_2.oid)
                                                                Filter: (((typname)::text = 'xml'::text) AND ((typtype = 'd'::"char") OR is_table_type(typrelid)))
                                                                ->  Bitmap Index Scan on pg_type_typname_nsp_index  (cost=0.00..38.97 rows=262 width=0)
                                                                      Index Cond: (typnamespace = s_2.oid)
                                                    ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_2  (cost=0.15..8.17 rows=1 width=64)
                                                          Index Cond: (nspname = base_2.nspname)
                                                          Filter: ((dbid)::oid = '5'::oid)
                                              ->  Function Scan on translate_pg_type_to_tsql tsql_type_name_2  (cost=0.00..0.01 rows=1 width=0)
                                                    Filter: (tsql_type_name IS NULL)
                                        ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name_1  (cost=0.00..0.01 rows=1 width=0)
                                  ->  Function Scan on is_table_type is_tbl_type  (cost=0.25..0.26 rows=1 width=0)
```


**Plan with this PR changes**

```
Query Text: select  
 tbl.name AS [Name] 
 , tbl.schema_id AS [Schema] 
 , CAST(ISNULL((select top 1 1 from all_columns as clmns join sys.types as usrt on usrt.user_type_id = clmns.user_type_id where clmns.object_id = tbl.object_id and usrt.name = N'xml'), 0) AS bit) AS [HasXmlData] 
from sys.tables AS tbl
Hash Semi Join  (cost=1883.89..15902.45 rows=4 width=37)
  Hash Cond: (t.relnamespace = base.oid)
  ->  Seq Scan on pg_class t  (cost=1870.63..2063.28 rows=43 width=72)
        Filter: ((NOT (hashed SubPlan 4)) AND has_schema_privilege(relnamespace, 'USAGE'::text COLLATE "default") AND has_table_privilege(oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER'::text COLLATE "default") AND (relkind = 'r'::"char") AND (relpersistence = ANY ('{p,u,t}'::"char"[])))
        SubPlan 4
          ->  Hash Join  (cost=187.68..1858.18 rows=4978 width=4)
                Hash Cond: (dep_1.objid = pt_1.typrelid)
                ->  Seq Scan on pg_depend dep_1  (cost=0.00..1488.71 rows=4244 width=4)
                      Filter: (deptype = 'i'::"char")
                ->  Hash  (cost=187.39..187.39 rows=23 width=8)
                      ->  Nested Loop  (cost=13.54..187.39 rows=23 width=8)
                            ->  Hash Semi Join  (cost=13.26..101.29 rows=114 width=4)
                                  Hash Cond: (pt_1.typnamespace = base_4.oid)
                                  ->  Seq Scan on pg_type pt_1  (cost=0.00..84.06 rows=1028 width=8)
                                        Filter: (typtype = 'c'::"char")
                                  ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                                        ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                              ->  Seq Scan on pg_namespace base_4  (cost=0.00..1.06 rows=1 width=68)
                                                    Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                              ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_4  (cost=0.15..8.17 rows=1 width=64)
                                                    Index Cond: (nspname = base_4.nspname)
                                                    Filter: ((dbid)::oid = '5'::oid)
                            ->  Index Scan using pg_class_oid_index on pg_class pc_1  (cost=0.28..0.76 rows=1 width=4)
                                  Index Cond: (oid = pt_1.typrelid)
                                  Filter: (relkind = 'r'::"char")
  ->  Hash  (cost=13.25..13.25 rows=1 width=4)
        ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
              ->  Seq Scan on pg_namespace base  (cost=0.00..1.06 rows=1 width=68)
                    Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
              ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext  (cost=0.15..8.17 rows=1 width=64)
                    Index Cond: (nspname = base.nspname)
                    Filter: ((dbid)::oid = '5'::oid)
  SubPlan 3
    ->  Limit  (cost=31.52..3456.43 rows=1 width=4)
          ->  Nested Loop  (cost=31.52..3456.43 rows=1 width=4)
                Join Filter: (t_1.oid = t_2.oid)
                ->  Nested Loop  (cost=13.82..209.20 rows=1 width=8)
                      ->  Nested Loop  (cost=13.82..209.18 rows=1 width=12)
                            ->  Nested Loop  (cost=13.82..209.15 rows=1 width=12)
                                  ->  Nested Loop  (cost=13.54..208.84 rows=1 width=4)
                                        Join Filter: has_column_privilege(((quote_ident((s.nspname)::text) || '.'::text COLLATE "default") || quote_ident((c.relname)::text)), (a.attname)::text, 'SELECT,INSERT,UPDATE,REFERENCES'::text COLLATE "default")
                                        ->  Nested Loop  (cost=13.25..188.11 rows=1 width=132)
                                              Join Filter: (c.relnamespace = s.oid)
                                              ->  Seq Scan on pg_class c  (cost=0.00..173.78 rows=1 width=72)
                                                    Filter: (((oid)::integer = (t.oid)::integer) AND (relkind = ANY ('{r,v,m,f,p}'::"char"[])))
                                              ->  Seq Scan on pg_namespace s  (cost=13.25..14.32 rows=1 width=68)
                                                    Filter: (has_schema_privilege(oid, 'USAGE'::text COLLATE "default") AND ((hashed SubPlan 1) OR (nspname = 'sys'::pg_catalog.name)))
                                                    SubPlan 1
                                                      ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                            ->  Seq Scan on pg_namespace base_1  (cost=0.00..1.06 rows=1 width=68)
                                                                  Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                            ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_1  (cost=0.15..8.17 rows=1 width=64)
                                                                  Index Cond: (nspname = base_1.nspname)
                                                                  Filter: ((dbid)::oid = '5'::oid)
                                        ->  Index Scan using pg_attribute_relid_attnum_index on pg_attribute a  (cost=0.29..20.55 rows=6 width=78)
                                              Index Cond: ((attrelid = c.oid) AND (attnum > 0))
                                              Filter: (NOT attisdropped)
                                  ->  Index Scan using pg_type_oid_index on pg_type t_1  (cost=0.28..0.31 rows=1 width=8)
                                        Index Cond: (oid = a.atttypid)
                            ->  Function Scan on translate_pg_type_to_tsql tsql_type_name  (cost=0.00..0.01 rows=1 width=0)
                      ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name  (cost=0.00..0.01 rows=1 width=0)
                ->  Append  (cost=17.69..3212.33 rows=1551 width=146)
                      CTE type_code_list
                        ->  HashAggregate  (cost=15.00..17.00 rows=200 width=64)
                              Group Key: babelfish_typecode_list.pg_typname, babelfish_typecode_list.tsql_typname
                              ->  Function Scan on babelfish_typecode_list  (cost=0.00..10.00 rows=1000 width=64)
                      ->  Nested Loop  (cost=0.69..14.42 rows=1 width=146)
                            ->  Nested Loop Left Join  (cost=0.69..13.63 rows=1 width=111)
                                  ->  Nested Loop  (cost=0.41..13.03 rows=1 width=51)
                                        ->  Nested Loop  (cost=0.28..12.83 rows=1 width=51)
                                              ->  CTE Scan on type_code_list ti  (cost=0.00..4.50 rows=1 width=64)
                                                    Filter: ((tsql_type_name IS NOT NULL) AND (tsql_type_name = 'xml'::text))
                                              ->  Index Scan using pg_type_typname_nsp_index on pg_type t_2  (cost=0.28..8.32 rows=1 width=83)
                                                    Index Cond: (typname = (ti.pg_type_name)::text)
                                                    Filter: pg_type_is_visible(oid)
                                        ->  Index Scan using pg_namespace_oid_index on pg_namespace s_1  (cost=0.13..0.17 rows=1 width=4)
                                              Index Cond: (oid = t_2.typnamespace)
                                              Filter: ((nspname = 'pg_catalog'::pg_catalog.name) OR (nspname = 'sys'::pg_catalog.name))
                                  ->  Index Scan using pg_collation_oid_index on pg_collation c_1  (cost=0.28..0.60 rows=1 width=68)
                                        Index Cond: (oid = t_2.typcollation)
                            ->  Function Scan on name default_collation_name  (cost=0.01..0.01 rows=1 width=64)
                      ->  Hash Right Join  (cost=1899.21..3157.65 rows=1550 width=146)
                            Hash Cond: (pt.typrelid = t_3.typrelid)
                            Filter: ((t_3.typtype = 'd'::"char") OR (pt.typrelid IS NOT NULL))
                            ->  Merge Join  (cost=1784.81..1823.33 rows=4978 width=4)
                                  Merge Cond: (pt.typrelid = dep.objid)
                                  ->  Sort  (cost=105.18..105.47 rows=114 width=4)
                                        Sort Key: pt.typrelid
                                        ->  Hash Semi Join  (cost=13.26..101.29 rows=114 width=4)
                                              Hash Cond: (pt.typnamespace = base_3.oid)
                                              ->  Seq Scan on pg_type pt  (cost=0.00..84.06 rows=1028 width=8)
                                                    Filter: (typtype = 'c'::"char")
                                              ->  Hash  (cost=13.25..13.25 rows=1 width=4)
                                                    ->  Nested Loop  (cost=0.15..13.25 rows=1 width=4)
                                                          ->  Seq Scan on pg_namespace base_3  (cost=0.00..1.06 rows=1 width=68)
                                                                Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                          ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_3  (cost=0.15..8.17 rows=1 width=64)
                                                                Index Cond: (nspname = base_3.nspname)
                                                                Filter: ((dbid)::oid = '5'::oid)
                                  ->  Sort  (cost=1679.62..1680.75 rows=452 width=8)
                                        Sort Key: dep.objid
                                        ->  Hash Join  (cost=159.82..1659.69 rows=452 width=8)
                                              Hash Cond: (dep.objid = pc.oid)
                                              ->  Seq Scan on pg_depend dep  (cost=0.00..1488.71 rows=4244 width=4)
                                                    Filter: (deptype = 'i'::"char")
                                              ->  Hash  (cost=150.19..150.19 rows=771 width=4)
                                                    ->  Seq Scan on pg_class pc  (cost=0.00..150.19 rows=771 width=4)
                                                          Filter: (relkind = 'r'::"char")
                            ->  Hash  (cost=114.39..114.39 rows=1 width=248)
                                  ->  Nested Loop  (cost=91.53..114.39 rows=1 width=248)
                                        ->  Nested Loop  (cost=91.53..114.37 rows=1 width=216)
                                              ->  Function Scan on name default_collation_name_1  (cost=0.01..0.01 rows=1 width=64)
                                              ->  Nested Loop Left Join  (cost=91.53..114.35 rows=1 width=152)
                                                    ->  Nested Loop  (cost=91.25..108.04 rows=1 width=92)
                                                          ->  Hash Right Join  (cost=91.10..95.86 rows=1 width=156)
                                                                Hash Cond: (ti_1.pg_type_name = t_3.typname)
                                                                Filter: (ti_1.tsql_type_name IS NULL)
                                                                ->  CTE Scan on type_code_list ti_1  (cost=0.00..4.00 rows=200 width=64)
                                                                ->  Hash  (cost=91.09..91.09 rows=1 width=156)
                                                                      ->  Nested Loop  (cost=0.00..91.09 rows=1 width=156)
                                                                            Join Filter: (t_3.typnamespace = base_2.oid)
                                                                            ->  Seq Scan on pg_namespace base_2  (cost=0.00..1.06 rows=1 width=68)
                                                                                  Filter: (nspname <> ALL ('{information_schema,pg_catalog,pg_toast,sys,public}'::pg_catalog.name[]))
                                                                            ->  Seq Scan on pg_type t_3  (cost=0.00..89.88 rows=12 width=92)
                                                                                  Filter: ((typname)::text = 'xml'::text)
                                                          ->  Index Scan using babelfish_namespace_ext_pkey on babelfish_namespace_ext ext_2  (cost=0.15..8.17 rows=1 width=64)
                                                                Index Cond: (nspname = base_2.nspname)
                                                                Filter: ((dbid)::oid = '5'::oid)
                                                    ->  Index Scan using pg_collation_oid_index on pg_collation c_2  (cost=0.28..6.29 rows=1 width=68)
                                                          Index Cond: (oid = t_3.typcollation)
                                        ->  Function Scan on translate_pg_type_to_tsql tsql_base_type_name_1  (cost=0.00..0.01 rows=1 width=32)
```